fs/move.py

```diff
@@ -134,9 +138,37 @@ def move_dir(
         preserve_time (bool): If `True`, try to preserve mtime of the
             resources (defaults to `False`).
 
+    Raises:
+        fs.errors.ResourceNotFound: if ``src_path`` does not exist on `src_fs`
+        fs.errors.DirectoryExpected: if ``src_path`` or one of its
+            ancestors is not a directory.
+        fs.errors.IllegalDestination: when moving a folder into itself
+
     """
     with manage_fs(src_fs, writeable=True) as _src_fs:
         with manage_fs(dst_fs, writeable=True, create=True) as _dst_fs:
+            if not _src_fs.exists(src_path):
+                raise ResourceNotFound(src_path)
+            if not _src_fs.isdir(src_path):
+                raise DirectoryExpected(src_path)
+
+            # if both filesystems have a syspath we use `shutil.move` to move the folder
+            if _src_fs.hassyspath(src_path) and _dst_fs.hassyspath(dst_path):
+                src_syspath = _src_fs.getsyspath(src_path)
+                dst_syspath = _dst_fs.getsyspath(dst_path)
+                # recheck if the move operation is legal using the syspaths
+                if isbase(src_syspath, dst_syspath):
+                    raise IllegalDestination(dst_path)
+                with _src_fs.lock(), _dst_fs.lock():
+                    with convert_os_errors("move_dir", dst_path, directory=True):
+                        if _dst_fs.exists(dst_path):
+                            os.rmdir(dst_syspath)
```

Comment:
Perhaps we don't need / want this `os.rmdir`?
Looking at the non-optimised "standard copy then delete" code, it looks like perhaps if `dst_path` already exists, _and_ contains existing contents, then the existing contents will be preserved after the `move_dir` call? :shrug: 
Certainly something that it's probably worth adding an extra test for :wink: